xayn-ai/src/coi/system.rs

```diff
@@ -214,6 +223,160 @@ impl CoiSystem {
+                .insert_axis(axis)
+        }
+
+        /// Gets the index of the maximum element.
+        fn argmax<I, F>(iter: I) -> Ix
+        where
+            I: IntoIterator<Item = F>,
+            F: Borrow<f32>,
+        {
+            iter.into_iter()
+                .enumerate()
+                .reduce(|(arg, max), (index, element)| {
+                    if element.borrow() > max.borrow() {
+                        (index, element)
+                    } else {
+                        (arg, max)
+                    }
+                })
+                .map(|(arg, _)| arg)
+                .unwrap(/* at least one key phrase is selected */)
+        }
+
+        let candidates = candidates
+            .iter()
+            .filter_map(|words| {
+                (!coi.key_phrases().contains(words))
+                    .then(|| {
+                        smbert(words)
+                            .ok()
+                            .and_then(|point| KeyPhrase::new(words, point).ok())
+                    })
+                    .flatten()
+            })
+            .collect::<BTreeSet<_>>();
+
+        let len = coi.key_phrases().len() + candidates.len();
+        let max_key_phrases = self.config.max_key_phrases.min(len);
+        if max_key_phrases == 0 {
+            coi.swap_key_phrases(BTreeSet::default());
+            return;
+        }
+
+        let similarity = pairwise_cosine_similarity(
+            coi.key_phrases()
+                .iter()
+                .chain(candidates.iter())
+                .map(|key_phrase| key_phrase.point().view())
+                .chain(once(coi.point().view())),
+        )
+        .slice_move(s![..len, ..]);
+        debug_assert!(similarity.iter().copied().all(f32::is_finite));
+
+        let min = reduce_without_diag(similarity.view(), Axis(0), f32::min, identity);
+        let max = reduce_without_diag(similarity.view(), Axis(0), f32::max, identity);
+        let normalized = (&similarity - &min) / (max - min);
+        let mean = reduce_without_diag(
+            normalized.view(),
+            Axis(0),
+            |reduced, element| reduced + element,
+            |reduced| reduced / (len - 1) as f32,
+        );
+        let std_dev = reduce_without_diag(
+            &normalized - &mean,
+            Axis(0),
+            |reduced, element| reduced + element.powi(2),
+            |reduced| (reduced / (len - 1) as f32).sqrt(),
+        );
+        let normalized = (normalized - mean) / std_dev + 0.5;
+        let normalized = normalized
+            .mapv_into(|normalized| normalized.is_finite().then(|| normalized).unwrap_or(0.5));
+        debug_assert!(normalized.iter().copied().all(f32::is_finite));
+
+        let candidate = argmax(normalized.slice(s![.., len]));
+        let mut selected = vec![false; len];
+        selected[candidate] = true;
+        for _ in 0..max_key_phrases - 1 {
+            let candidate = argmax(selected.iter().zip(normalized.rows()).map(
```

Comment:
If that is correct I think we can precompute the max 3 elements of each candidate and work with that.

```rust

let candidate = argmax(...);
let selected = HashSet::new();
selected.insert(candidate);
let candidates: HashMap<Ix, Vec<(Ix, f32)> = /*max 3 from each candidate*/;
for _ in 0..max_key_pharases -1 {
  let candidate = candidates.iter().map(|(idx, vs)| 
    (idx, vs.iter().filter(|(ix, _| !selected.contains(ix)).max_by_key(|p| p.1).unwrap())
  ).max_by_key(|p| p.1);
  selected.insert(candidate);
  candidates.remove(candidate);
}
```

I think there is also a way where we we have a vector of `(v, row_ix, col_ix)` (just the top `max_key_phrases` for each `row_ix`), we order it and then we take the first `max_key_phrases` with unique index. But not sure.

Reply:
> From my understanding we try to select a new index

that's right, the next index from the yet unselected is picked

> that have the biggest `self.config.gamma * ... * max`

it's important to note that this is a sum of products, not just a product: `gamma * similarity + (1-gamma) * max`

> where max is computed removing the already selected phrases.

on the contrary, max is computed only on the selected phrases, not on the removed.

> If that is correct I think we can precompute the max 3 elements of each candidate and work with that.

this won't work, because at any iteration of the outer loop those 3 elements (or any other fixed number) might be still unselected, so they don't get used in the `max` computation. we would need to hold all indices and values in the `candidates: HashMap<Ix, Vec<(Ix, f32)>`, which would be possible, but looks much less efficient than just holding them in a matrix (the normalized similarities) accompanied by a single indexing vector (the selected indices) like it's currently done.

Reply:
> it's important to note that this is a sum of products, not just a product: `gamma * similarity + (1-gamma) * max`
> 

I used `* ... *` but didn't want to imply that was only a product just that I was omitting part of it. Did I missing something else?

> > where max is computed removing the already selected phrases.
> 
> on the contrary, max is computed only on the selected phrases, not on the removed.

Ah! I completely misunderstood that. I think I saw that in the first case we return `MIN` if selected to "ignore" it and I assumed the same also for the rest of the code 🙈. Yeah what I though is wrong.

Reply:
> Did I missing something else?

i don't think so